Implement a batch import UI for punches

totten opened the pull request:
This defines a screen `civicrm/timetrack/import` with a big textarea. It accepts a variant of the `!pi -s` notation and displays a preview of the errors and normalized data as you edit.

![screen shot 2019-02-27 at 12 46 47 am](https://user-images.githubusercontent.com/1336047/53477306-422fa680-3a29-11e9-86b0-62bb18adb1b1.png)

totten committed:
Update civix boilerplate

totten committed:
Timetrackpunchlist API - Implement helpers for previewing and importing batches

totten committed:
Import phpunit boilerplate

totten committed:
Implement civicrm/timetrack/import (#9)

This addresses the import scenario for #9.  It includes live-update, so that
you can fix syntax errors and data-errors while watching the preview of the
proposed import.

mlutfy commented:
Neat. It's kind of overkill, but the code is light and it's very efficient.

Would it be possible to remove the `!pi -s` part? The `-s` is mandatory in this case anyway, and assuming the `!pi -s` prefix would make it a bit more user-friendly?

totten committed:
Timetrackpunchlist.preview - Make the "!pi -s" optional

totten committed:
ImportCtrl.html - Fix typo in format doc

totten committed:
Timetrackpunchlist.preview - Accept day-of-week as a sanity check

I find it more useable to keep notes in a format with the day-of-week.
Helps to jog the memory when making the notes, and it's a sanity-check to
make sure I got the date right.

totten committed:
Timetrackpunchlist - When importing, be more forgiving about extra spaces

totten committed:
Timetrackpunchlist - Accept 'hr' and 'min' as units (in addition to 'h', 'H', 'm', 'M')

totten committed:
Timetrackpunchlist - If time is omitted, use noon

totten committed:
Timetrackpunchlist - Simplify/flatten regex groupings

totten committed:
Timetrackpunchlist - More intuitive error for empty comments

totten committed:
ImportCtrl.html - Show time as optional. Format duration in hours.

totten commented:
@bgm Yeah, I had imagined it being nice to match the format of the bot exactly, but it's true that the `!pi -s` doesn't buy a whole lot for someone importing. I've updated to relax it. (The importer basically ignores `!pi -s` if it's present.)

Additionally, I tested locally with my current backlog (~100 punches over ~6 weeks). Unsurprisingly, there were some inconsistencies (e.g. `ct/d8` vs `ct/drupal8`; e.g. blank comments), but that was easy to spot+fix with the importer's messages. I also revised a couple rules, though, to accommodate some things which occurred frequently:

* The most common date format in my backlog was "`DOW MM/DD`"... it seems that I think more naturally on a day-to-day basis in terms of day-of-week. The DOW helps me when looking for major omissions, and it's a sanity-check to make sure I got the dates right. So the importer now accepts either `YYYY-MM-DD` or `DOW MM/DD`. If the DOW doesn't match, it'll complain.
* In nearly all cases, I recorded duration but not start-time. I think I originally tried start-times at first... but it made the data-entry feel fiddly (extra work; not a huge gain from my POV/use-case). Omitting it made it easier to handle some weird scenarios (like multitasking or like splitting time when multiple categories are applicable). Anyway, a start-time is required in the data-model, so I just fill-in blanks as `12:00`. (*I was worried it might not render well on `civicrm/timetrack/timeline`... but it does seem to render alright.*)

Anyway, the import succeeded with my current backlog on a local dev-site. :)

mlutfy commented:
@totten You rock! Thank you!
